enhancements/olm/olm-managed-operator-metrics.md

```diff
@@ -0,0 +1,149 @@
+---
+title: olm-integration-with-openshift-monitoring
+authors:
+  - "@awgreene"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2019-12-11
+last-updated: 2019-12-11
+status: implementable
+---
+
+# olm-integration-with-openshift-monitoring
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary and Motivation
+
+The primary purpose of this enhancement is to propose a series of changes that would enable [Operator-Lifecycle-Manager (OLM)](https://github.com/operator-framework/operator-lifecycle-manager) managed operators to report that metrics are available to the [OpenShift Monitoring](https://github.com/openshift/cluster-monitoring-operator) Prometheus Operator.
+
+First, OLM managed operators should be capable of packaging `ServiceMonitors` and `PrometheusRule` objects within their bundle to interact with generic [Prometheus Operators](https://github.com/coreos/prometheus-operator) on cluster.
```

Comment:
I think this is missing a key requirement, that all operators on the platform (whether they expose metrics or not) expose USE (usage, saturation, errors) consistently for low effort.  That's the pattern CVO uses, and I think we need that here - "if you write a good operator, you get USE for free".  This proposal is too focused on mechanical "how do we expose all metrics from the operators", which is ok, but "exposing metrics" isn't the goal, "exposing good metrics" is.

For instance, we can constrain the volume / cardinality limits by having OLM manage a service monitor that is a variant of the one they create that only allows specific labels.  We can also make prometheus handle fairness.  We can also have OLM expose usage (number of objects created within the CRDs created by this operator) and errors (is the operator responding to requests).

Reply:
Specifically I would expect this proposal to cover OLM automatically offering a minimal USE set of metrics that doesn't require the operator to expose any metrics, like the CVO does for "cluster_operator_conditions".   That is a guardrail / accessible on-ramp that allows everyone to take a step up.

Reply:
Good point, I agree we should have USE metrics for each operator, but those are not complete enough, we should also add RED metrics. Since errors are included in both then ->  utilisation, saturation, rate, errors, duration metrics.

Reply:
supposedly there is another EP that covers "default metrics for operators, reported by OLM on the operator's behalf" (similar to what the CVO does).  @shawn-hurley or @jwforres can you link it so we don't get sidetracked in this EP?

Reply:
There was another epic, I don't believe that we have created an enhancement proposal for this. 

I think it would help to frame this conversation as "allow olm managed operators to expose metrics to the cluster monitoring". @awgreene can we update the title

Reply:
I have updated the title.